src/JitInterface/src/CorInfoImpl.cs

```diff
@@ -962,54 +962,121 @@ private void getGSCookie(IntPtr* pCookieVal, IntPtr** ppCookieVal)
-
-                object resultUninstantiated = methodILUninstantiated.GetObject((int)pResolvedToken.token);
+            object result = ResolveTokenInScope(methodIL, typeOrMethodContext, pResolvedToken.token);
 
-                if (resultUninstantiated is MethodDesc)
-                {
-                    return ((MethodDesc)resultUninstantiated).InstantiateSignature(typeInstantiation, methodInstantiation);
-                }
-                else if (resultUninstantiated is FieldDesc)
+            if (result is MethodDesc method)
+            {
+                if (method.IsSharedByGenericInstantiations)
                 {
-                    return ((FieldDesc)resultUninstantiated).InstantiateSignature(typeInstantiation, methodInstantiation);
+                    MethodDesc sharedMethod = methodIL.OwningMethod.GetSharedRuntimeFormMethodTarget();
+                    result = ResolveTokenWithSubstitution(methodIL, pResolvedToken.token, sharedMethod.OwningType.Instantiation, sharedMethod.Instantiation);
+                    Debug.Assert(((MethodDesc)result).IsRuntimeDeterminedExactMethod);
                 }
-                else
+            }
+            else if (result is FieldDesc field)
+            {
+                if (field.OwningType.IsCanonicalSubtype(CanonicalFormKind.Any))
                 {
-                    TypeDesc result = ((TypeDesc)resultUninstantiated).InstantiateSignature(typeInstantiation, methodInstantiation);
-                    if (pResolvedToken.tokenType == CorInfoTokenKind.CORINFO_TOKENKIND_Newarr)
-                        result = result.MakeArrayType();
-                    return result;
+                    MethodDesc sharedMethod = methodIL.OwningMethod.GetSharedRuntimeFormMethodTarget();
+                    result = ResolveTokenWithSubstitution(methodIL, pResolvedToken.token, sharedMethod.OwningType.Instantiation, sharedMethod.Instantiation);
+                    Debug.Assert(((FieldDesc)result).OwningType.IsRuntimeDeterminedSubtype);
                 }
             }
             else
             {
-                object result = methodIL.GetObject((int)pResolvedToken.token);
-                if (pResolvedToken.tokenType == CorInfoTokenKind.CORINFO_TOKENKIND_Newarr)
-                    return ((TypeDesc)result).MakeArrayType();
+                TypeDesc type = (TypeDesc)result;
+                if (type.IsCanonicalSubtype(CanonicalFormKind.Any))
+                {
+                    MethodDesc sharedMethod = methodIL.OwningMethod.GetSharedRuntimeFormMethodTarget();
+                    result = ResolveTokenWithSubstitution(methodIL, pResolvedToken.token, sharedMethod.OwningType.Instantiation, sharedMethod.Instantiation);
+                    Debug.Assert(((TypeDesc)result).IsRuntimeDeterminedSubtype ||
+                        ((TypeDesc)methodIL.GetObject((int)pResolvedToken.token)).IsCanonicalDefinitionType(CanonicalFormKind.Any));
```

Comment:
Just for my education - I apologize but I have a bit of a hard time to get my head around this check. How exactly it differs from the above condition 

<pre>
type.IsCanonicalSubtype(CanonicalFormKind.Any
</pre>

Would it be possible to simplify this by changing this to

<pre>
type.Context.IsCanonicalDefinitionType(type, CanonicalFormKind.Any)
</pre>

?

Reply:
I managed to mess up the assert - the assert was more loose than I wanted it to be. We want to call the `GetObject` on the uninstantiated `MethodIL` and check for `__Canon` there.

It's supposed to only match literal `typeof(__Canon)`, not `typeof(T)` in a shared method body. Thanks for bringing my attention to it!